### PR TITLE
[gpu/enhance] Registering OpenCL kernels at cl_context

### PR DESCRIPTION
**Pull request by s-debadri**
Changes made in this PR:

- Added `registerClKernel` function to register custom OpenCl kernels as well as in-house kernels.
- Exposed `CommandQueueManager` and `ContextManager` instances of `cl_context`.
- Used hash map to track created `Kernel` objects.
- Modified `createLayer` API signature to accept compute engine information.
- Modified blas kernels to remove `cl_context` related dependencies.
- Removed kernel management logic from `layer_context`.
- Temporarily commented out `cl_layers` build and initialization to avoid build failure.

> To Do: Remove dependency of `layer_context` for all `cl_layers`. This can be done in subsequent PR.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Debadri Samaddar <s.debadri@samsung.com>

### CODE REVIEW

File: nntrainer/cl_context.h [hunk 1]
```diff
@@ -183,12 +191,33 @@ class ClContext {
...
+                                            std::string kernel_name);
+
+  /**
+   * @brief create OpenCl kernel
+   * @param kernel_string reference of implementation string
+   * @param kernel_name reference of kernel_name
+   * @param kernel_ptr_ reference of shared_ptr of Kernel
+   * @return true if successful, false otherwise
+   */
+  bool clCreateKernel(std::string &kernel_string, std::string &kernel_name,
```
Comment (EunjuYang) [inline nntrainer/cl_context.h:210]: As I understood, it is not recommended call `clCreateKernel` directly, but should be call `registerClKernel`. Therefore, could we consider altering 'clCreateKernel' into a private member function?

**Review by EunjuYang — COMMENTED** [review]
It seems beneficial to update the CL functions so they do not require `RunLayerContext` as a parameter! I believe this PR may be linked to Issue #2723. Would you mind briefly sharing your plans regarding this matter in the issue? I feel that it would assist the reviewers in understanding our intended course of action.

**Comment by s-debadri** [comment]
> It seems beneficial to update the CL functions so they do not require `RunLayerContext` as a parameter! I believe this PR may be linked to Issue #2723. Would you mind briefly sharing your plans regarding this matter in the issue? I feel that it would assist the reviewers in understanding our intended course of action.

Added the course of action in Issee #2723

File: nntrainer/cl_context.h (hunk 1 above)
Comment (s-debadri) [reply nntrainer/cl_context.h]: Updated the commit with this change

**Comment by jijoongmoon** [comment]
I thought we might need another kernel context to handle the kernels more efficiently. But embracing the kernels in cl context is also fine for me. However, I suggest handling the blas kernels as a kind of default kernels in NNTrainer if opencl option is enabled so that, we can create kernels at initialize time. These kernels are used and they might be better managed in the same directory. so.. make separate files (let's says blas_kernels.h) and save the blas-related kernel strings in this file and register them when the cl context is created. Then we can use them in cl tensors without clKernelCreate call.

**Review by jijoongmoon — APPROVED** [review]
Overall, LGTM

**Comment by s-debadri** [comment]
> I thought we might need another kernel context to handle the kernels more efficiently. But embracing the kernels in cl context is also fine for me. However, I suggest handling the blas kernels as a kind of default kernels in NNTrainer if opencl option is enabled so that, we can create kernels at initialize time. These kernels are used and they might be better managed in the same directory. so.. make separate files (let's says blas_kernels.h) and save the blas-related kernel strings in this file and register them when the cl context is created. Then we can use them in cl tensors without clKernelCreate call.

Got it. Shall I make the changes related to default kernel registration in this same PR or different?

**Pull request merged**